### ISSUE TITLE
Adding small REPL for development purpose

### ISSUE THREAD

**Issue by navidR** [Microsoft/ChakraCore#1619]
I am trying to read and understand ChakraCore architecture. In process of doing that, when I want to change or fix small part of ChakraCore runtime, I have to run tests or something, which is unnecessary for most of the time.

I am developing small REPL inside of ChakraCore project which could be helpful to many people (I think) so every time I am trying something new or messing with something I can run and test it simply inside Visual Studio without going through all the pain of testing and other stuff. It is for my own usage right now, but we can work on it to make it something better and ultimately adding it to main repo.

What is your opinion to add small REPL like subproject (or it could be "-repl" flag for ch.exe) ?

**Comment by dilijev** [comment]
I talked about adding this before, but since it's easy enough to invoke `ch.exe test.js` most people didn't view this as valuable. I, for one, would welcome a `ch.exe -repl` mode, or even simply `ch.exe -e "console.log(3+5)"` (`-e` for execute)

**Comment by navidR** [comment]
That -e is nice for one line evaluation.So simple to implement too.

About node-chakra , that idea crossed my mind  too, but that doesn't improve the cycle of dev/compile/run because after compiling ChakraCore you should every time embed the newly compiled chakracore inside nodejs (unless I am missing something) so nodejs-chakra would not improve situation in this specific case.

**Comment by obastemur** [comment]
>> ChakraCore you should every time embed the newly compiled chakracore inside nodejs (unless I am missing something) so nodejs-chakra would not improve situation in this specific case.

At this very moment, there are some minor difference between the node-chakracore and chakracore repos. (xplat API differences) but we will have them sorted out very soon. Once they share the same code base, you might clone your chakracore fork into node-chakracore, play with it + make changes and compile node-chakracore. There you have your changes in place.

Yet, `-e` and `-pe` options sounds like a good improvement though.

```
ch -pe "1.245.toFixed(2)"
```

or similarly;

```
ch -e "print(1.245.toFixed(2)"
```

Honestly, I find `-pe` option is similarly useful.

**Comment by dilijev** [comment]
@navidR are you planning to implement this? I'll assign it to you for now :)

**Comment by navidR** [comment]
@dilijev yes, but it may take a little while (because of my exams). Thanks

**Comment by dilijev** [comment]
@navidR Just checking. Good luck with exams. Thanks!

**Comment by mathiasbynens** [comment]
IMHO `-e` and REPL are two different features. One does not imply that the other isn’t needed anymore. I’d like to have both!

Ideally `ch` would support piping as well, e.g. `echo 'print(1+2)' | ch`.

Note that the `--repl` flag doesn’t seem necessary — `ch` could default to the REPL when invoked without getting a file passed to it.

**Comment by dilijev** [comment]
@mathiasbynens - Agreed. On the issue of time investment, though, the `-e` flag would be minimal effort and add a lot of value, whereas the REPL will take a much more significant engineering effort. We would still very much welcome PRs for that, though!

**Comment by navidR** [comment]
Sorry for the delay, I was so busy in recent month, I do have a personal implementation for "-e", let me polish it a little bit, I will submit a patch. I think pipe idea is very good too.

But repl do need a little bit restructuring of code in Ch.

**Comment by dilijev** [comment]
@navidR `p1 | p2` sends stdout of `p1` to stdin of `p2`. In order to get piping to ch to work, you would just have to implement handling for stdin. stdin processing could take an entire script followed by EOT (`^D` / `\x04`) or EOF and basically behave like `-e`, OR (possibly later) behave like a REPL except not to display a prompt until stdin is empty of all input lines (thus effectively behaving the same as before, with an interactive option). This is basically how most programs with a REPL seem to behave anyway.

**Comment by dilijev** [comment]
Hi @navidR - Gentle reminder, and just wondering, have you made any progress on this?

**Comment by navidR** [comment]
Sorry for the delay. Yes. I did implement it but I only could add this kind of flag (after a little bit of tweaking in ConsoleFlagParser ) to ch : 

> ch.exe -eval:"Console.Write('Hello World')" 

What do you think ? Do you think this kind of flag is good enough ?
The reason I hesitate is that personally I prefer to have standard unix style flag. But ch only have this kind of flag. (those other flags you see when you type help, all redirected to ChakraCore itself). 

I haven't spare time to add pipe to ch.exe, yet.

**Comment by dilijev** [comment]
@navidR You mean you couldn't add a `-e` or you couldn't add `--eval`. I think the general feeling is that we would like to keep things consistent across the board. We don't explicitly support unix-style short flags (in the sense that you cannot chain flags under a single `-` like in `rm-rf` which is a shorthand for `rm -r -f`. We also don't make a distinction between single-character and multiple-character flags. However, for every supported flag, IIRC we allow any of the three prefixes `/` (Windows-only), `-`, or `--`. The first is for backwards compat on Windows but we prefer not to use it (it's not supported on xplat anyway). By convention we prefer `-` over `--` just for general consistency.

You're right that most flags are passed through to ChakraCore itself. However, you can add logic directly to `ch` to parse flags ahead of time for `ch`-specific behavior like this. See the [implementation of `ch -version`](https://github.com/Microsoft/ChakraCore/pull/2490/files#diff-9aa417ac81c06da693354e50d1edd449R850) for example.

Or were you talking about needing to use a colon to associate the parameter with the flag? I think you could do something like reading the next parameter if you detect the flag `-e` or `-eval` and then advancing by 2. We already do this in a few places.

**Comment by navidR** [comment]
> You're right that most flags are passed through to ChakraCore itself. However, you can add logic directly to ch to parse flags ahead of time for ch-specific behavior like this. See the implementation of ch -version for example.

I knew, the problem with this approach is we cannot send flags to ChakraCore engine itself while we using  it in eval mode, I want to be able : 
```
./ch -eval="WScript.Echo('Hello Chakra')" --flags-for-chakracore --flags-for-chakracore
```

I have added eval flag to host flags in Ch. It works for this syntax right now 
```
./ch -eval:"WScript.Echo('Hello Chakra')" --flags-for-chakracore --flags-for-chakracore
```
I am trying to find a way to make it work for 
```
./ch -eval="WScript.Echo('Hello Chakra')" --flags-for-chakracore --flags-for-chakracore
```
too.

BTW I tried hard to only use existing code instead of adding haphazard flag handling, and Ch's existing flag handling only handles long flags like "-eval" . What do you think? do you think it is enough to have only one of "-eval" or "-e" or I should find a new workaround?

**Comment by dilijev** [comment]
@navidR Can you push that change to a branch and share a link to the branch so we can more easily evaluate.

As for alternatives of the same flag, you could do something like having an alternative boolean for each one, have the presence of either set the same boolean, etc.

I would make the `([-]-eval|-e)` flag handled in ch itself, using the same code pattern that parses out `([-]-version|-v)`, load the script string instead of loading a script file (which is also done by ch), and then `ch` passes all remaining flags to the underlying engine. I'm not really sure I understand the problem you're encountering.

**Comment by dilijev** [comment]
FWIW there's a tool called eshost that might serve the purpose of the `--eval` you're trying to add here (also doesn't have a REPL). https://gist.github.com/dilijev/d8f2fd58551360409b2ba3f50af144fd

That said I still think it would be good to have this functionality built into the `ch` host we publish.

**Comment by navidR** [comment]
https://github.com/navidR/ChakraCore/tree/eval
e.g. 
```
[navid@workstation ChakraCore]$ ./out/Debug/ch -eval:"function hello(){var regex = /blah/;WScript.Echo('blah: ' + regex.blah);regex.blah = 1;WScript.Echo('blah: ' + regex.blah);}
hello();hello();"
function hello(){var regex = /blah/;WScript.Echo('blah: ' + regex.blah);regex.blah = 1;WScript.Echo('blah: ' + regex.blah);}hello();hello();
blah: undefined
blah: 1
blah: undefined
blah: 1
[navid@workstation ChakraCore]$ cat test/Object/regex.baseline
blah: undefined
blah: 1
blah: undefined
blah: 1
[navid@workstation ChakraCore]$
```

> IMO --eval doesn't belong in the HostConfigFlagList, but rather it belongs to the same area of code as the --version etc. It's possible others will disagree, however I think that will be the easiest way to do this.

I disagree, I think "-version" and other stuff should have used Ch existed class for flag handling.

**Comment by dilijev** [comment]
> I disagree, I think "-version" and other stuff should have used Ch existed class for flag handling.

@navidR You're probably right that this change belongs there, although as you've seen it doesn't support flag aliases. As for `--version` the reason we couldn't put it there is because we wanted it to be available in release builds as well without having to refactor the existing class (which is debug/test only) to be enabled and to make a distinction between test and release flags.

**Comment by 853419196** [comment]
In order to facilitate the development, I wrote a REPL compatible with ChakraCore in JavaScript.
https://github.com/853419196/JavaScript_REPL
